pymc/tests/distributions/test_multivariate.py

```diff
@@ -1367,6 +1391,100 @@ def test_issue_3706(self):
+            v = pm.ZeroSumNormal("v", dims=dims, shape=shape, zerosum_axes=zerosum_axes)
+            s = pm.sample(10, chains=1, tune=100)
+
+        # to test forward graph
+        random_samples = pm.draw(
+            v,
+            draws=10,
+        )
+
+        assert s.posterior.v.shape == (1, 10, len(COORDS["regions"]), len(COORDS["answers"]))
+
+        if not isinstance(zerosum_axes, (list, tuple)):
+            zerosum_axes = [zerosum_axes]
+
+        if isinstance(zerosum_axes[0], str):
+            for ax in zerosum_axes:
+                for samples in [
+                    s.posterior.v.mean(dim=ax),
+                    random_samples.mean(axis=dims.index(ax) + 1),
+                ]:
+                    assert np.isclose(
+                        samples, 0
+                    ).all(), f"{ax} is a zerosum_axis but is not summing to 0 across all samples."
+
+            nonzero_axes = list(set(dims).difference(zerosum_axes))
+            if nonzero_axes:
+                for ax in nonzero_axes:
+                    for samples in [
+                        s.posterior.v.mean(dim=ax),
+                        random_samples.mean(axis=dims.index(ax) + 1),
+                    ]:
+                        assert not np.isclose(
+                            samples, 0
+                        ).all(), f"{ax} is not a zerosum_axis, but is nonetheless summing to 0 across all samples."
+
+        else:
+            for ax in zerosum_axes:
+                if ax < 0:
+                    assert np.isclose(
+                        s.posterior.v.mean(axis=ax), 0
+                    ).all(), f"{ax} is a zerosum_axis but is not summing to 0 across all samples."
+                else:
+                    ax = ax + 2  # because 'chain' and 'draw' are added as new axes after sampling
+                    assert np.isclose(
+                        s.posterior.v.mean(axis=ax), 0
+                    ).all(), f"{ax} is a zerosum_axis but is not summing to 0 across all samples."
+
+    @pytest.mark.parametrize(
+        "dims, zerosum_axes",
+        [
+            (("regions", "answers"), 2),
+            (("regions", "answers"), (0, -2)),
+        ],
+    )
+    def test_zsn_fail_axis(self, dims, zerosum_axes):
+        if isinstance(zerosum_axes, (list, tuple)):
+            with pytest.raises(ValueError, match="repeated axis"):
+                with pm.Model(coords=COORDS) as m:
+                    _ = pm.ZeroSumNormal("v", dims=dims, zerosum_axes=zerosum_axes)
+        else:
+            with pytest.raises(AxisError, match="out of bounds"):
+                with pm.Model(coords=COORDS) as m:
+                    _ = pm.ZeroSumNormal("v", dims=dims, zerosum_axes=zerosum_axes)
+
+    def test_zsn_change_dist_size(self):
+        base_dist = pm.ZeroSumNormal.dist(shape=(4, 9))
+
+        new_dist = change_dist_size(base_dist, new_size=(5, 3), expand=False)
+        assert new_dist.eval().shape == (5, 3)
```

Comment:
You should also test that once you change size and draw some random numbers from it, your zero sum axes actually sum to zero

Reply:
Just tested this (see updated `test_zsn_change_dist_size`), and it seems to work 🥳

Reply:
You need to test after change dist size is called, not before